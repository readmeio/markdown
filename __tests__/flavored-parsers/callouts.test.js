import { mdast } from '../../index';

describe('Parse RDMD Callouts', () => {
  it('renders an info callout', () => {
    const text = `
> ℹ️ Info Callout
>
> Lorem ipsum dolor  sit amet consectetur adipisicing elit.`;

    expect(mdast(text)).toMatchSnapshot();
  });

  it('supports a default theme', () => {
    const text = `
> 🥇 Themeless
>
> Lorem ipsum dolor sit amet consectetur adipisicing elit.`;

    expect(mdast(text)).toMatchSnapshot();
  });

  it('parses a callout with no title', () => {
    const text = `
> ℹ️
>
> Lorem ipsum dolor  sit amet consectetur adipisicing elit.`;

    expect(mdast(text)).toMatchSnapshot();
  });

  describe('edge cases', () => {
    it('renders html inside a callout', () => {
      const text = `
> ℹ️ Info Callout
>
> <span>With html!</span>
`;

      expect(mdast(text)).toMatchSnapshot();
    });

    it('does not allow trailing spaces after the icon with no title', () => {
      const text = `
> 🛑 
> Compact headings must be followed by two line breaks before the following block.`;

      expect(mdast(text)).toMatchSnapshot();
    });
  });

  it('requires a space between the icon and title', () => {
    const text = `
> ℹ️Info Callout
>
> Lorem ipsum dolor  sit amet consectetur adipisicing elit.`;

    expect(mdast(text)).toMatchSnapshot();
  });

  it('allows nested callouts', () => {
    const text = `
- list item
  > ℹ️ Info Callout
  >
  > Lorem ipsum dolor  sit amet consectetur adipisicing elit.`;

    expect(mdast(text)).toMatchSnapshot();
  });

  it('does not require a line break between the title and the body', () => {
    const text = `
> 💁 Undocumented Behavior
> Lorem ipsum dolor  sit amet consectetur adipisicing elit.`;

    expect(mdast(text)).toMatchSnapshot();
  });
});

describe('emoji modifier support', () => {
  const emojis = ['📘', '🚧', '⚠️', '👍', '✅', '❗', '❗️', '🛑', '⁉️', '‼️', 'ℹ️', '⚠'];

  emojis.forEach(emoji => {
    it(`render a callout for ${emoji}`, () => {
      const text = `
> ${emoji}
>
> Lorem ipsum dolor sit amet consectetur adipisicing elit.`;
      const ast = mdast(text);
      expect(ast).toMatchSnapshot();
    });
  });
});
