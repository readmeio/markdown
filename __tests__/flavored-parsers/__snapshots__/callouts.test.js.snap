// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse RDMD Callouts allows nested callouts 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "list item",
                },
              ],
              "type": "paragraph",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Info Callout",
                    },
                  ],
                  "type": "paragraph",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "data": Object {
                "hName": "rdme-callout",
                "hProperties": Object {
                  "icon": "‚ÑπÔ∏è",
                  "theme": "info",
                  "title": "Info Callout",
                  "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
                },
              },
              "type": "rdme-callout",
            },
          ],
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts does not require a line break between the title and the body 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Undocumented Behavior",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üíÅ",
          "theme": "default",
          "title": "Undocumented Behavior",
          "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts edge cases does not allow trailing spaces after the icon with no title 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Compact headings must be followed by two line breaks before the following block.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üõë",
          "theme": "error",
          "title": "",
          "value": "Compact headings must be followed by two line breaks before the following block.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts edge cases renders html inside a callout 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Info Callout",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "html",
              "value": "<span>",
            },
            Object {
              "type": "text",
              "value": "With html!",
            },
            Object {
              "type": "html",
              "value": "</span>",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ÑπÔ∏è",
          "theme": "info",
          "title": "Info Callout",
          "value": "<span>With html!</span>",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts parses a callout with no title 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ÑπÔ∏è",
          "theme": "info",
          "title": "",
          "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts renders an info callout 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Info Callout",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ÑπÔ∏è",
          "theme": "info",
          "title": "Info Callout",
          "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts requires a space between the icon and title 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "‚ÑπÔ∏èInfo Callout",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor  sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "blockquote",
    },
  ],
  "type": "root",
}
`;

exports[`Parse RDMD Callouts supports a default theme 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Themeless",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "ü•á",
          "theme": "default",
          "title": "Themeless",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ÄºÔ∏è 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ÄºÔ∏è",
          "theme": "error",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ÅâÔ∏è 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ÅâÔ∏è",
          "theme": "error",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ÑπÔ∏è 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ÑπÔ∏è",
          "theme": "info",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ö† 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ö†",
          "theme": "warn",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ö†Ô∏è 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ö†Ô∏è",
          "theme": "warn",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚úÖ 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚úÖ",
          "theme": "okay",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ùó 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ùó",
          "theme": "error",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for ‚ùóÔ∏è 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "‚ùóÔ∏è",
          "theme": "error",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for üëç 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üëç",
          "theme": "okay",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for üìò 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üìò",
          "theme": "info",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for üöß 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üöß",
          "theme": "warn",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`emoji modifier support render a callout for üõë 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üõë",
          "theme": "error",
          "title": "",
          "value": "Lorem ipsum dolor sit amet consectetur adipisicing elit.",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;
